### ISSUE TITLE
UTF-8 generation of java files

### ISSUE THREAD

**Issue by damnms** [edeandrea/xjc-generation-gradle-plugin#5]
Is there any way to create UTF-8 encoded source files from xjc?

I get a lot of errors like:
`invoice400\BalanceType.java:4: error: unmappable character (0xC4) for encoding UTF-8`

**Comment by edeandrea** [comment]
Hi @damnms do you have a simple example that would show/isolate this problem? I'm not sure how to try to even reproduce it or add it to the test suite to make sure that it doesn't happen again.

**Comment by damnms** [comment]
Unfortunately not, its a big fat old project. When i work with my linux laptop, i do not have any problem with that. Now i do homeoffice and on my windows machine i am facing those problems.
I found that: https://stackoverflow.com/questions/22777386/how-to-make-cxf-xjc-plugin-generate-sources-in-utf-8

**Comment by damnms** [comment]
I guess the xjc binary takes different encoding, depending on the OS. And for windows, it is this wierd windows-1252, but i have everything else for UTF-8. But thats just a guess...

**Comment by edeandrea** [comment]
Yeah I've been looking as well. If you can't submit an entire example can you provide a few things that will help me troubleshoot?

1. The `xjc` configuration dependencies from your `dependencies` block
1. The `xjcGeneration` block from your build file (you can whitewash/sanitize it if you wish to remove anything specific about your organization)
1. If you can, the wsdl/xsd file that you are generating from. If you can't share the entire wsdl/schema, just the `<?xml` declaration as well as the element in the xsd/wsdl containing the character that is making things blow up
1. Whats the default locale of the machine that is throwing the errors? US English? Something else?

**Comment by damnms** [comment]
```
plugins {
    id 'com.github.edeandrea.xjc-generation' version '1.1'
}

ext {
    jaxbVersion = '2.2.11'
}

sourceSets {
    main.java.srcDirs += 'build/generated-sources/main/xjc'
    test.java.srcDirs += 'build/generated-sources/main/xjc'
}

xjcGeneration {


    schemas {

        schema40 {
            schemaRootDir = 'src/main/resources/xsd'
            schemaFile = 'generalInvoiceRequest_400.xsd'
            javaPackageName = 'ch.forum_datenaustausch.invoice400'
        }

        schema41 {
            schemaRootDir = 'src/main/resources/xsd'
            schemaFile = 'generalInvoiceRequest_410.xsd'
            javaPackageName = 'ch.forum_datenaustausch.invoice410'
        }

        commonTypes {
            schemaRootDir = 'src/main/resources/xsd'
            schemaFile = 'commonTypes.xsd'
            javaPackageName = 'com.mycompany.xsd.eclaim.common'
        }

        invoice {
            schemaRootDir = 'src/main/resources/xsd'
            schemaFile = 'invoice.xsd'
            javaPackageName = 'com.mycompany.xsd.eclaim.invoice'
        }

        syriusRuleDef {
            schemaRootDir = 'src/main/resources/xsd'
            schemaFile = 'syriusRuleDef.xsd'
        }
    }
}

dependencies {

    xjc "javax.xml.bind:jaxb-api:$jaxbVersion"
    xjc "com.sun.xml.bind:jaxb-impl:$jaxbVersion"
    xjc "com.sun.xml.bind:jaxb-xjc:$jaxbVersion"
    xjc "com.sun.xml.bind:jaxb-core:$jaxbVersion"
    xjc 'javax.activation:activation:1.1.1'

    // JAX-B dependencies for JDK 9+
    implementation "jakarta.xml.bind:jakarta.xml.bind-api:2.3.2"
    implementation "org.glassfish.jaxb:jaxb-runtime:2.3.2"

    compile 'joda-time:joda-time:2.3'
    compile 'org.springframework.data:spring-data-jpa:1.4.3.RELEASE'
    compile 'commons-io:commons-io:2.4'
    compile 'org.springframework.batch:spring-batch-core:3.0.1.RELEASE'
    compile 'com.fasterxml.jackson.core:jackson-annotations:2.1.0'
    compile 'cglib:cglib:2.0.2'
    compile 'commons-collections:commons-collections:3.2.1'
    testCompile 'junit:junit:4.12'
}
```

the files are available in the internet: https://www.forum-datenaustausch.ch/de/
unfortunately, i can not publish them here because github forbids .xsd - seems dangerous. funny that .docx is allowed 👎

**Comment by damnms** [comment]
```
plugins {
    id 'com.github.edeandrea.xjc-generation' version '1.1'
}

ext {
    jaxbVersion = '2.2.11'
}

sourceSets {
    main.java.srcDirs += 'build/generated-sources/main/xjc'
    test.java.srcDirs += 'build/generated-sources/main/xjc'
}

xjcGeneration {


    schemas {

        schema40 {
            schemaRootDir = 'src/main/resources/xsd'
            schemaFile = 'generalInvoiceRequest_400.xsd'
            javaPackageName = 'ch.forum_datenaustausch.invoice400'
        }

        schema41 {
            schemaRootDir = 'src/main/resources/xsd'
            schemaFile = 'generalInvoiceRequest_410.xsd'
            javaPackageName = 'ch.forum_datenaustausch.invoice410'
        }

        commonTypes {
            schemaRootDir = 'src/main/resources/xsd'
            schemaFile = 'commonTypes.xsd'
            javaPackageName = 'com.mycompany.xsd.eclaim.common'
        }

        invoice {
            schemaRootDir = 'src/main/resources/xsd'
            schemaFile = 'invoice.xsd'
            javaPackageName = 'com.mycompany.xsd.eclaim.invoice'
        }

        someRuleDef {
            schemaRootDir = 'src/main/resources/xsd'
            schemaFile = 'someRuleDef.xsd'
        }
    }
}

dependencies {

    xjc "javax.xml.bind:jaxb-api:$jaxbVersion"
    xjc "com.sun.xml.bind:jaxb-impl:$jaxbVersion"
    xjc "com.sun.xml.bind:jaxb-xjc:$jaxbVersion"
    xjc "com.sun.xml.bind:jaxb-core:$jaxbVersion"
    xjc 'javax.activation:activation:1.1.1'

    // JAX-B dependencies for JDK 9+
    implementation "jakarta.xml.bind:jakarta.xml.bind-api:2.3.2"
    implementation "org.glassfish.jaxb:jaxb-runtime:2.3.2"

    compile 'joda-time:joda-time:2.3'
    compile 'org.springframework.data:spring-data-jpa:1.4.3.RELEASE'
    compile 'commons-io:commons-io:2.4'
    compile 'org.springframework.batch:spring-batch-core:3.0.1.RELEASE'
    compile 'com.fasterxml.jackson.core:jackson-annotations:2.1.0'
    compile 'cglib:cglib:2.0.2'
    compile 'commons-collections:commons-collections:3.2.1'
    testCompile 'junit:junit:4.12'
}
```

the files are available in the internet: https://www.forum-datenaustausch.ch/de/
unfortunately, i can not publish them here because github forbids .xsd - seems dangerous. funny that .docx is allowed 👎

**Comment by damnms** [comment]
```
plugins {
    id 'com.github.edeandrea.xjc-generation' version '1.1'
}

ext {
    jaxbVersion = '2.2.11'
}

sourceSets {
    main.java.srcDirs += 'build/generated-sources/main/xjc'
    test.java.srcDirs += 'build/generated-sources/main/xjc'
}

xjcGeneration {


    schemas {

        schema40 {
            schemaRootDir = 'src/main/resources/xsd'
            schemaFile = 'generalInvoiceRequest_400.xsd'
            javaPackageName = 'ch.forum_datenaustausch.invoice400'
        }

        schema41 {
            schemaRootDir = 'src/main/resources/xsd'
            schemaFile = 'generalInvoiceRequest_410.xsd'
            javaPackageName = 'ch.forum_datenaustausch.invoice410'
        }

        commonTypes {
            schemaRootDir = 'src/main/resources/xsd'
            schemaFile = 'commonTypes.xsd'
            javaPackageName = 'com.mycompany.xsd.common'
        }

        invoice {
            schemaRootDir = 'src/main/resources/xsd'
            schemaFile = 'invoice.xsd'
            javaPackageName = 'com.mycompany.xsd.invoice'
        }

        someRuleDef {
            schemaRootDir = 'src/main/resources/xsd'
            schemaFile = 'someRuleDef.xsd'
        }
    }
}

dependencies {

    xjc "javax.xml.bind:jaxb-api:$jaxbVersion"
    xjc "com.sun.xml.bind:jaxb-impl:$jaxbVersion"
    xjc "com.sun.xml.bind:jaxb-xjc:$jaxbVersion"
    xjc "com.sun.xml.bind:jaxb-core:$jaxbVersion"
    xjc 'javax.activation:activation:1.1.1'

    // JAX-B dependencies for JDK 9+
    implementation "jakarta.xml.bind:jakarta.xml.bind-api:2.3.2"
    implementation "org.glassfish.jaxb:jaxb-runtime:2.3.2"

    compile 'joda-time:joda-time:2.3'
    compile 'org.springframework.data:spring-data-jpa:1.4.3.RELEASE'
    compile 'commons-io:commons-io:2.4'
    compile 'org.springframework.batch:spring-batch-core:3.0.1.RELEASE'
    compile 'com.fasterxml.jackson.core:jackson-annotations:2.1.0'
    compile 'cglib:cglib:2.0.2'
    compile 'commons-collections:commons-collections:3.2.1'
    testCompile 'junit:junit:4.12'
}
```

the files are available in the internet: https://www.forum-datenaustausch.ch/de/
unfortunately, i can not publish them here because github forbids .xsd - seems dangerous. funny that .docx is allowed 👎

**Comment by edeandrea** [comment]
Thanks I will take a look. Just a minor thing I noticed...

```groovy
sourceSets {
    main.java.srcDirs += 'build/generated-sources/main/xjc'
    test.java.srcDirs += 'build/generated-sources/main/xjc'
}
```

You shouldn't need to have this in your build file. The generated output should automatically be linked to your sources. If they aren't then thats a bug :)

**Comment by damnms** [comment]
the problem here is with intellij - intellij recognizes those files only as java sources if i add this 2 lines, otherwise those folders are like regular folders with txt files in it

**Comment by edeandrea** [comment]
Really? Thats odd. I use IntelliJ myself and don't see it like that. The plugin [is doing that](https://github.com/edeandrea/xjc-generation-gradle-plugin/blob/e92e38591b3675eb4debbd09e82ccb0754c0ba00/src/main/kotlin/com/github/edeandrea/xjcplugin/plugin/XjcPlugin.kt#L94).

**Comment by damnms** [comment]
i will check that tomorrow, maybe thats "dead" code 
at least i thought that was the problem with those lines, but its a couple of weeks since i made that, maybe i was wrong ;)

**Comment by edeandrea** [comment]
I think what I'm going to try to do is implement this solution as I would implement #4 . Specify a way for specific options to be passed into xjc directly. I can't really change the default since that will not be backwards compatible, but provide options to allow expressing what the encoding should be and to be able to set a default encoding if not present.

I'll let you know once I have something.

**Comment by edeandrea** [comment]
According to the [XJC Ant Task docs](https://javaee.github.io/jaxb-v2/doc/user-guide/ch04.html#tools-xjc-ant-task) seems the default encoding is whatever the system encoding is

   > Set the encoding name for generated sources, such as EUC-JP or UTF-8. If it is not specified, the platform default encoding is used.

**Comment by edeandrea** [comment]
Hi @damnms the ability for this has been implemented in version 1.2 of the plugin. I am not able to apply UTF-8 encoding by default since that would break backwards compatibility, but you can apply it for all of your schemas by doing

```groovy
xjcGeneration {
  defaultAdditionalXjcOptions = ['encoding': 'UTF-8']
}
```

or on a schema-by-schema basis by doing

```groovy
xjcGeneration {
  schemas {
      schema40 {
        schemaRootDir = 'src/main/resources/xsd'
        schemaFile = 'generalInvoiceRequest_400.xsd'
        javaPackageName = 'ch.forum_datenaustausch.invoice400'
        additionalXjcOptions = ['encoding': 'UTF-8']
    }
  }
}
```

**Comment by damnms** [comment]
thanks! i will try this tomorrow
